include/mixer.h

```diff
@@ -480,4 +481,63 @@ void MIXER_SetReverbPreset(const ReverbPreset new_preset);
 ChorusPreset MIXER_GetChorusPreset();
 void MIXER_SetChorusPreset(const ChorusPreset new_preset);
 
+// Generic callback used for audio devices which generate audio on the main thread
+// These devices produce audio on the main thread and consume on the mixer thread
+// This callback is the consumer part
+template <class DeviceType, class AudioType, bool stereo, bool signeddata, bool nativeorder>
+inline void MIXER_PullFromQueueCallback(const int frames_requested, DeviceType* device)
+{
+	// Currently only handles mono sound (output_queue is a primitive type and frames == samples)
+	// Special case for AudioType == AudioFrame (stereo floating-point sound)
+	static_assert((!stereo) || std::is_same<AudioType, AudioFrame>::value);
+
+	// AudioFrame type is always stereo
+	static_assert(stereo || (!std::is_same<AudioType, AudioFrame>::value));
+
+	size_t clamped_frames = check_cast<size_t>(frames_requested);
+	if (MIXER_InFastForwardMode()) {
+		// Special case, normally only hit when using the fast-forward hotkey (Alt + F12)
+		// We need a very large buffer to compensate or it results in static
+		device->output_queue.Resize(iceil(device->channel->GetFramesPerBlock() * 100.0f));
```

Comment:
Is `100.0f` just a number out of thin air? Shouldn't this buffer's size be adjusted dynamically on demand? We can't really anticipate how "fast" fast-forward mode can be, e.g., on future CPUs. Will it generate static at some point when it goes "too fast"?

Reply:
If a CPU is running at faster than 100x real-time then yeah maybe you get static.  Don't think it's worth complicating the code to dynamically size this.  You can basically scale this number up arbitrarily if it ever becomes a problem.  The only downside is increased latency but audio is chipmunk'd in this path anyway so who cares.

Reply:
Sounds reasonable. Could you explain this in a comment? Seeing `* 100.0f` there casually is very misleading. First I thought, are we converting to some fractional percentage to real percentage? Maybe `MaxExpectedFastForwardFactor = 100.0f` would be enough.

Reply:
Sure, extracted to a constant and added a comment.